File: src/main/java/com/alxnns1/mobhunter/entity/monsters/EntityMHHerbivore.java
package com.alxnns1.mobhunter.entity.monsters;

import com.alxnns1.mobhunter.entity.IScaledMob;
import com.alxnns1.mobhunter.init.MHItems;
import com.alxnns1.mobhunter.reference.Config;
import net.minecraft.entity.IEntityLivingData;
import net.minecraft.entity.SharedMonsterAttributes;
import net.minecraft.entity.ai.*;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.passive.EntityAnimal;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.datasync.DataParameter;
import net.minecraft.network.datasync.DataSerializers;
import net.minecraft.network.datasync.EntityDataManager;
import net.minecraft.world.DifficultyInstance;
import net.minecraft.world.World;

/**
 * Created by Mark on 28/04/2016.
 */
public abstract class EntityMHHerbivore extends EntityAnimal implements IScaledMob
{
    private static final String KEY_SCALE = "scale";
    private static final DataParameter<Float> ENTITY_SCALE = EntityDataManager.createKey(EntityMHHerbivore.class, DataSerializers.FLOAT);
    private static float scaleMax; //= 1.24f;
    private static float scaleMin; //= 0.79f;
    private double baseHealth = 5d;
    private double baseSpeed = 0.1d;
    private double baseKnockback = 0.2d;
    private Item breedItem;

    public EntityMHHerbivore(World world)
    {
        this(world, Config.scaleMin, Config.scaleMax, Items.WHEAT);
    }

    public EntityMHHerbivore(World world, float minScale, float maxScale)
    {
        this(world, minScale, maxScale, Items.WHEAT);
    }

    public EntityMHHerbivore(World world, Item temptFood)
    {
        this(world, Config.scaleMin, Config.scaleMax, temptFood);
    }

    public EntityMHHerbivore(World world, float minScale, float maxScale, Item temptFood)
    {
        super(world);
        this.setSize(0.9F, 1.3F); //Same as cow
        scaleMin = minScale;
        scaleMax = maxScale;
        breedItem = temptFood;
        this.tasks.addTask(0, new EntityAISwimming(this));
        this.tasks.addTask(1, new EntityAIPanic(this, 2.0D));
        this.tasks.addTask(2, new EntityAIMate(this, 1.0D));
        this.tasks.addTask(3, new EntityAITempt(this, 1.25D, temptFood, false));
        this.tasks.addTask(4, new EntityAIFollowParent(this, 1.25D));
        this.tasks.addTask(5, new EntityAIWander(this, 1.0D));
        this.tasks.addTask(6, new EntityAIWatchClosest(this, EntityPlayer.class, 6.0F));
        this.tasks.addTask(7, new EntityAILookIdle(this));
    }

    protected void entityInit()
    {
        super.entityInit();
        //Creates the datawatcher object to save the entity scale in
        this.dataManager.register(ENTITY_SCALE, 1.0f);
    }

    //Same as pig
    protected void applyEntityAttributes()
    {
        super.applyEntityAttributes();
        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(10.0D);
        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.25D);
    }

    protected Item getDropItem()
    {
        return MHItems.itemHerbivoreDrop;
    }

    /**
     * Checks if the parameter is an item which this animal can be fed to breed it
     */
    public boolean isBreedingItem(ItemStack stack)
    {
        return stack == null ? false : stack.getItem() == breedItem;
    }

    /**
     * Called only once on an entity when first time spawned, via egg, mob spawner, natural spawning etc, but not called
     * when entity is reloaded from nbt. Mainly used for initializing attributes and inventory
     */
    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, IEntityLivingData livingdata)
    {
        float scale = (this.rand.nextFloat() * (scaleMax - scaleMin)) + scaleMin;
        this.setEntityScale(scale);
        return super.onInitialSpawn(difficulty, livingdata);
    }

    private void setEntityScale(float scale)
    {
        //Gets the datawatcher value for the entity scale
        this.dataManager.set(ENTITY_SCALE, scale);
        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue((double) Math.round(baseHealth * scale));
        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(baseSpeed * scale);
        this.getEntityAttribute(SharedMonsterAttributes.KNOCKBACK_RESISTANCE).setBaseValue(baseKnockback * scale);
        this.setHealth(this.getMaxHealth());
    }

    public float getScale()
    {
        return this.dataManager.get(ENTITY_SCALE);
    }

    public void writeEntityToNBT(NBTTagCompound tagCompound)
    {
        super.writeEntityToNBT(tagCompound);
        tagCompound.setFloat(KEY_SCALE, getScale());
    }

    public void readEntityFromNBT(NBTTagCompound tagCompund)
    {
        super.readEntityFromNBT(tagCompund);
        setEntityScale(tagCompund.getFloat(KEY_SCALE));
    }

    protected void setBaseHealth(int health)
    {
        baseHealth = (double) health;
    }

    protected void setBaseSpeed(double speed)
    {
        baseSpeed = speed;
    }

    protected void setBaseKnockback(double knockback)
    {
        baseKnockback = knockback;
    }

    /**
     * Drops a single item
     */
    public EntityItem dropSingleItem(Item item)
    {
        return dropSingleItem(item, 0);
    }

    /**
     * Drops a single item
     */
    public EntityItem dropSingleItem(Item item, int meta)
    {
        return entityDropItem(new ItemStack(item, 1, meta), 0);
    }
}

File: src/main/java/alxnns1/mobhunter/entity/monsters/EntityMHHerbivore.kt
package alxnns1.mobhunter.entity.monsters

import alxnns1.mobhunter.entity.IScaledMob
import alxnns1.mobhunter.init.MHItems
import alxnns1.mobhunter.reference.MHConfig
import net.minecraft.entity.IEntityLivingData
import net.minecraft.entity.SharedMonsterAttributes
import net.minecraft.entity.ai.*
import net.minecraft.entity.passive.EntityAnimal
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.datasync.DataSerializers
import net.minecraft.network.datasync.EntityDataManager
import net.minecraft.world.DifficultyInstance
import net.minecraft.world.World

abstract class EntityMHHerbivore @JvmOverloads constructor(
        world: World,
        minScale: Double = MHConfig.scaleMin,
        maxScale: Double = MHConfig.scaleMax,
        private val breedItem: Item = Items.WHEAT
) : EntityAnimal(world), IScaledMob {

    companion object {
        private val KEY_SCALE = "scale"
        private val ENTITY_SCALE = EntityDataManager.createKey(EntityMHHerbivore::class.java, DataSerializers.FLOAT)
        private var scaleMax = 0.0
        private var scaleMin = 0.0
    }

    var baseHealth = 5.0
    var baseSpeed = 0.1
    var baseKnockBack = 0.2

    constructor(world: World, temptFood: Item) : this(world, MHConfig.scaleMin, MHConfig.scaleMax, temptFood)

    init {
        setSize(0.9f, 1.3f)
        scaleMin = minScale
        scaleMax = maxScale
        tasks.addTask(0, EntityAISwimming(this))
        tasks.addTask(1, EntityAIPanic(this, 2.0))
        tasks.addTask(2, EntityAIMate(this, 1.0))
        tasks.addTask(3, EntityAITempt(this, 1.25, breedItem, false))
        tasks.addTask(4, EntityAIFollowParent(this, 1.25))
        tasks.addTask(5, EntityAIWander(this, 1.0))
        tasks.addTask(6, EntityAIWatchClosest(this, EntityPlayer::class.java, 6.0f))
        tasks.addTask(7, EntityAILookIdle(this))
    }

    override fun entityInit() {
        super.entityInit()
        dataManager.register(ENTITY_SCALE, 1.0f)
    }

    override fun applyEntityAttributes() {
        super.applyEntityAttributes()
        getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).baseValue = 10.0
        getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).baseValue = 0.25
    }

    override fun getDropItem() = MHItems.itemHerbivoreDrop!!

    override fun isBreedingItem(stack: ItemStack) = !stack.isEmpty && stack.item == breedItem

    override fun onInitialSpawn(difficulty: DifficultyInstance?, livingData: IEntityLivingData?): IEntityLivingData? {
        setEntityScale(rand.nextFloat() * (scaleMax - scaleMin) + scaleMin)
        return super.onInitialSpawn(difficulty, livingData)
    }

    private fun setEntityScale(scale: Double) {
        dataManager.set(ENTITY_SCALE, scale.toFloat())
        getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).baseValue = Math.round(baseHealth * scale).toDouble()
        getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).baseValue = baseSpeed * scale
        getEntityAttribute(SharedMonsterAttributes.KNOCKBACK_RESISTANCE).baseValue = baseKnockBack * scale
        health = maxHealth
    }

    override fun getScale() = dataManager.get(ENTITY_SCALE)!!

    override fun writeEntityToNBT(tagCompound: NBTTagCompound) {
        super.writeEntityToNBT(tagCompound)
        tagCompound.setFloat(KEY_SCALE, scale)
    }

    override fun readEntityFromNBT(tagCompound: NBTTagCompound) {
        super.readEntityFromNBT(tagCompound)
        setEntityScale(tagCompound.getFloat(KEY_SCALE).toDouble())
    }

    @JvmOverloads
    fun dropSingleItem(item: Item, meta: Int = 0) = entityDropItem(ItemStack(item, 1, meta), 0f)
}
